keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -117,16 +122,14 @@ When reconciling an object `Foo` a Controller must:
 1. Send the trace context stored in `Foo` in the http request context for all API requests. See [Tracing API Requests](#tracing-api-requests)
 1. Store the trace context of `Foo` in object `Bar` when updating the Spec of `Bar`. See [Propagating Context Through Objects](#propagating-context-through-objects)
 1. Export a span around work that attempts to drive the actual state of an object towards its desired state
-1. Replace the trace context of `Foo` when updating `Foo`'s status to the desired state
+1. Remove the trace context of `Foo` when updating `Foo`'s status to the desired state
```

Comment:
> Remove the trace context of `Foo` when updating `Foo`'s status to the desired state

is it possible? b/c we're actually filtering annotations changes upon updating status-subresouce. or we will have to do another non-status update request which suspiciously triggers another unnecessary around of recenciliation..

Reply:
It worked when I wrote my demo ~6 months ago.  I had initially thought I wouldn't be able to update object annotations as part of a status subresource update, but in practice it appears to work.  Are there changes introduced in the past ~2 releases that would prevent this from working now?

Reply:
Updating metadata fields in status updates was accidentally allowed for some resources (like pod and node), but is not for newly introduced resources (volumeattachments, custom resources, etc). Don't count on being able to do that uniformly.

Reply:
Great catch @yue9944882 ! I think that having a decidated non-status update request is a blocker for different reasons (including scalability).

Reply:
I'm still thinking about how to resolve this.  A few options come to mind:
1. Don't "end" traces in components by removing the trace context.  Traces now contain all spans collected between object writes that update the trace context.
2. Only "end" traces for objects that allow metadata updates with status updates.
3. Generally allow updating annotations as part of subresource updates.
4. Add a field in the object that can be updated as part of all updates.

I suspect the ability to update metadata fields was disallowed for good reason, so I suspect 3 isn't a viable option.  4 seems like it would be disallowed for the same reasons as 3...  @liggitt can you provide any pointers on why metadata updates were disallowed?

While "ending" the trace by removing the context is useful for end-users, it is also a source of complexity for the implementation of tracing in controllers.  Maybe 1 wouldn't be entirely bad. Thoughts @logicalhan @DirectXMan12 ?

Reply:
see https://github.com/kubernetes/kubernetes/issues/45539

Reply:
@liggitt Is https://github.com/kubernetes/kubernetes/issues/20978#issuecomment-324220862 still accurate?  I.E. if we added this as an explicit field in object meta (option #4), we could allow updates via subresources?

Reply:
I haven't kept up with the current state of this proposal, so I'm not sure... a few questions:
* is the ability to mess with trace contexts considered security-related?
* does the API requester have control over updates to that field, or are the values placed into that field controlled by the API rest handler?

Reply:
> is the ability to mess with trace contexts considered security-related?

Yes.  We ideally only want actors with either (1) the ability to write to the Spec or Status of an object to be able to update the trace context.  We could potentially narrow that to say that access to the /status subresource only allows clearing the field, not updating to a different value.

> does the API requester have control over updates to that field, or are the values placed into that field controlled by the API rest handler?

The API requester (e.g. users using kubectl, or controllers writing an object) have control over updates to the field.  This is needed in order to "propagate" a context they may already be acting under.

Reply:
>> is the ability to mess with trace contexts considered security-related?
>
> Yes. We ideally only want actors with either the ability to write to the Spec or Status of an object to be able to update the trace context. We could potentially narrow that to say that access to the /status subresource only allows clearing the field, not updating to a different value.

Would that enforcement be baked into the API server? Applying strong validation and specific update rules for specific annotations with the unstructured annotations map for all objects is a bit odd.

>> does the API requester have control over updates to that field, or are the values placed into that field controlled by the API rest handler?
>
> The API requester (e.g. users using kubectl, or controllers writing an object) have control over updates to the field. This is needed in order to "propagate" a context they may already be acting under.

Is there any concept of which contexts a given actor is associated with and allowed to add? What are the implications of an actor adding an unrelated context to an object?

Reply:
> Would that enforcement be baked into the API server? Applying strong validation and specific update rules for specific annotations with the unstructured annotations map for all objects is a bit odd.

I probably wouldn't want to do strong validation for annotations.  At that point, I would assume this needs its own field in ObjectMeta, which is what I was exploring [above](https://github.com/kubernetes/enhancements/pull/1458#discussion_r399513983)

> Is there any concept of which contexts a given actor is associated with and allowed to add? What are the implications of an actor adding an unrelated context to an object?

Any writing actor can overwrite the trace context.  The plan is to use trace Links (which can cross from one trace to the next) to allow users to follow traces even after their context is overwritten.  The current proposal suggests the overwriting actor should do this, although that certainly won't protect against malicious actors.  With an explicit (non-annotation) field, we could actually have the API Server add the link, and not allow deletion of the trace context, only overwriting with a new valid context.  This would ensure the trace from an overwritten context still exists somewhere.  Users would just have to follow the link.

Reply:
I talked with @liggitt offline, and it the best option seems to be adding this as an explicit field in ObjectMeta.  I modified the proposal accordingly.